superbench/benchmarks/micro_benchmarks/directx_render_performance/GeometryHelper.h

```diff
@@ -0,0 +1,153 @@
+};
+
+// Create a random geometry data buffer.
+template <class T> std::unique_ptr<Geometry> CreateRandomGeometry(const UINT vertexNum, const UINT indexNum) {
+    static_assert(std::is_base_of<Vertex, T>::value, "T must be a Vertex or derived from Vertex");
+    std::unique_ptr<Geometry> geo = make_unique<Geometry>();
+    // Create the vertices.
+    // Allocate memory and reinterpret_cast it to Vertex array
+    geo->vertexData.reset(reinterpret_cast<Vertex *>(new T[vertexNum]));
+
+    // Fill in the random vertex data.
+    for (UINT i = 0; i < vertexNum; i++) {
+        // Here you need to reinterpret_cast it back to T for accessing/modifying
+        T &v = reinterpret_cast<T &>(geo->vertexData[i]);
+        v = T();
+    }
+
+    // Create the indices.
+    // Fill in the random index data.
+    for (UINT i = 0; i < indexNum; i++) {
+        geo->indexData.push_back(MathHelper::genRand2N_large(vertexNum));
+    }
+    geo->vertexNum = vertexNum;
+    geo->indexNum = indexNum;
+    geo->vertexByteStride = sizeof(T);
+    geo->vertexByteSize = sizeof(T) * vertexNum;
+    geo->indexByteSize = sizeof(std::uint16_t) * indexNum;
+    return geo;
+}
+
+// Helpter class to manage geometry data buffer on GPU.
+struct GeometryResource {
+    ComPtr<ID3DBlob> VertexBufferCPU = nullptr;
+    ComPtr<ID3DBlob> IndexBufferCPU = nullptr;
+
+    ComPtr<ID3D12Resource> VertexBufferGPU = nullptr;
+    ComPtr<ID3D12Resource> IndexBufferGPU = nullptr;
+
+    ComPtr<ID3D12Resource> VertexBufferUploader = nullptr;
+    ComPtr<ID3D12Resource> IndexBufferUploader = nullptr;
+
+    // Data about the buffers.
+    UINT VertexByteStride = 0;
+    UINT VertexBufferByteSize = 0;
+    DXGI_FORMAT IndexFormat = DXGI_FORMAT_R16_UINT;
+    UINT IndexBufferByteSize = 0;
+
+    D3D12_PRIMITIVE_TOPOLOGY PrimitiveType = D3D11_PRIMITIVE_TOPOLOGY_TRIANGLELIST;
+
+    // Target geometry.
+    std::unique_ptr<Geometry> geometry = nullptr;
+    UINT IndexCount = 0;
+    UINT StartIndexLocation = 0;
+    INT BaseVertexLocation = 0;
+
+    /*
+     * @brief Get the vertex buffer view.
+     */
+    D3D12_VERTEX_BUFFER_VIEW VertexBufferView() const {
+        D3D12_VERTEX_BUFFER_VIEW vbv;
+        vbv.BufferLocation = VertexBufferGPU->GetGPUVirtualAddress();
+        vbv.StrideInBytes = VertexByteStride;
+        vbv.SizeInBytes = VertexBufferByteSize;
+        return vbv;
+    }
+
+    /*
+     * @brief Get the index buffer view.
+     */
+    D3D12_INDEX_BUFFER_VIEW IndexBufferView() const {
+        D3D12_INDEX_BUFFER_VIEW ibv;
+        ibv.BufferLocation = IndexBufferGPU->GetGPUVirtualAddress();
+        ibv.Format = IndexFormat;
+        ibv.SizeInBytes = IndexBufferByteSize;
+        return ibv;
+    }
+
+    /*
+     * @brief Upload geometry data and set necessary information about the geometry.
+     */
+    void Create(ID3D12Device *device, ID3D12GraphicsCommandList *cmdList, std::unique_ptr<Geometry> &geoData) {
+        geometry = std::move(geoData);
```

Comment:
>   geometry = std::move(geoData); [](http://example.com/codeflow?start=5&length=33)

Looks like you don't need to save this variable "geometry" in this structure, so please re-consider it to see if we can remove it. :smile:

Reply:
yes, I removed it, I don't need to save this geometry at all, will not use it later